src/pin.ts

```diff
@@ -0,0 +1,339 @@
+import { flags, colors } from './deps.ts'
+import { readFile, writeFile, mkdir, copyFile, readdir } from 'node:fs/promises'
+import { existsSync } from 'node:fs'
+import { resolve, join, dirname, basename } from 'node:path'
+import process from 'node:process'
+
+/**
+ * Options for the pin command
+ */
+interface PinOptions {
+  overwrite?: boolean
+  'only-changed'?: boolean
+}
+
+/**
+ * Structure of chelonia.json configuration file
+ * This is ecosystem-agnostic and separate from package.json
+ */
+interface CheloniaConfig {
+  contracts: {
+    [contractName: string]: {
+      version: string
+      path: string // relative path to the contract file
+    }
+  }
+  // Future: can add other Chelonia-specific configuration here
+  // like deployment targets, signing keys, etc.
+}
+
+/**
+ * ContractPinner class handles pinning individual contracts to specific versions
+ *
+ * Key improvements over the old implementation:
+ * 1. Uses chelonia.json instead of package.json for ecosystem-agnostic config
+ * 2. Supports per-contract versioning instead of global versioning
+ * 3. Uses contracts/<name>/<version>/<files...> directory structure
+ * 4. Only updates the specific contract being pinned, not all contracts
+ * 5. Maintains compatibility with chel manifest and chel deploy commands
```

Comment:
This is an example of a comment that can be useful within a PR but which isn't as useful in the source code. For example, `Key improvements over the old implementation:` is not useful without context about what the 'old implementation' is, and the same goes for the 'instead of' comments. I'm not saying to remove the comment, but to reword it in a way that's useful to someone who is looking at this for the first time and may not have the background to know what this is referring to.